File: Experiments/GRoadHog/GRoadHog.cpp
//	Google Drive + excessive "driving" == G[oogle] RoadHog

// This program transforms any input file into a bunch of CSV files
// which can be uploaded at Google Drive for free (0 bytes of storage!) as "Google Docs".

//This version is written in C++17, as it works faster than Python-based prototype.

#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>

//Constants

const std::string ALPHABET {"0123456789abcdefghijklmnop"};
const unsigned int MAX_STRING {50000 / 2};
const unsigned int MAX_COLUMNS {3};
const unsigned int MAX_ROWS {122};

//Encode one byte into buffer
void encode_byte (char* buf, char byte) {
	//Get hexademical representation of the byte 
	snprintf(buf, 3, "%02x", (unsigned char) byte);
	//Apply the Caesar cipher
	buf[0]=ALPHABET[ALPHABET.find(buf[0]) + 10];
	buf[1]=ALPHABET[ALPHABET.find(buf[1]) + 10];
}

int main (int argc, char *argv[]) {
	//Wrong quantity of parameters
	if (argc < 2) {
		std::cout << "Please provide the name of source file." << std::endl;
		return 0;
	}
	//Buffer for encoding, one more byte for '\0' symbol
	char buf[3];
	//Open input file
	std::string path_to_file = argv[1];
	std::ifstream source(path_to_file, std::ios_base::binary);
	//Current index of CSV file block
	unsigned long long current_file {0};
	//Current length of the cell
	unsigned int current_length {0};
	//Current index of the column
	unsigned int current_column {0};
	//Current index of the row
	unsigned int current_row {0};
	while (true) {
		//Generate correct filename for current CSV block
		std::string filename {std::to_string(current_file)};
		filename.append(".csv");
		//Print it out
		std::cout << filename << std::endl;
		{
			//Create new CSV file
			std::ofstream destination(filename);
			while (true) {
				//Read one byte
				source.get(buf[0]);
				//EOF? Enough.
				if (!source) {
					return 0;
				}
				//Otherwise, proceed with encoding
				encode_byte(buf, buf[0]);
				//Write the ciphered symbols
				destination << buf[0] << buf[1];
				//Check if current length, column, row or the whole file is filled
				++current_length;
				if (current_length >= MAX_STRING) {
					current_length = 0;
					++current_column;
					if (current_column >= MAX_COLUMNS) {
						current_column = 0;
						++current_row;
						if (current_row >= MAX_ROWS) {
							current_row = 0;
							++current_file;
							//Go to the next CSV file block
							break;
						//Add 'new line' symbol if neccessary
						} else destination << '\n';
					//Add 'tab' symbol if neccessary
					}  else destination << '\t';
				}
			}
		}
	}
}

File: Experiments/GRoadHog/GRoadHog.py
#	Google Drive + excessive "driving" == G[oogle] RoadHog

# This program transforms any input file into a bunch of CSV files
# which can be uploaded at Google Drive for free (0 bytes of storage!) as "Google Docs".

import sys

#Name of the source file (from the command prompt)
SOURCE_FILE=sys.argv[1]

#Constants:
MAX_STRING=50000
MAX_COLUMNS=3
MAX_ROWS=122

#Constant-based constant values:
MAX_STRING2=int(MAX_STRING / 2)
MAX_COLUMNS2=MAX_COLUMNS - 1
MAX_ROWS2=MAX_ROWS - 1

#Store encoded data in CSV:
def WriteToCSV(f):
	#Current CSV file index
	CSVNum=0
	while True:
		with open(''.join((str(CSVNum), ".csv")), "w") as file_csv:
			for row in range(MAX_ROWS):
				for col in range(MAX_COLUMNS):
					for _ in range(MAX_STRING2):
						#Read one byte
						onebyte=f.read(1)
						#EOF? Halt.
						if not onebyte: return
						#Get a hex code of the byte
						hex_code='%02x' % ord(onebyte)
						#Apply the Caesar cipher (avoid treating symbols like numbers on Google side) 
						code="".join([chr(int(i, 16) + ord('a')) for i in hex_code])
						#Store the result in the CSV file block:
						file_csv.write(code)
					#Add 'tab' symbol if neccessary (end of cell)
					if col != MAX_COLUMNS2:
						file_csv.write("\t")
				#Add 'new line' symbol if neccessary (end of row)
				if row != MAX_ROWS2:
					file_csv.write("\n")
		#This file block is full, proceed to another 
		CSVNum+=1

def EncodeFile(filename):
	with open(filename,"rb") as f: WriteToCSV(f)

EncodeFile(SOURCE_FILE)